### DISCUSSION TITLE
Translating MonitorControl to Your Language!

### DISCUSSION THREAD

**Discussion by waydabber** [MonitorControl/MonitorControl#637] [General]
**Dear translators,**

thank you for translating MonitorControl to your native language! **Your contribution is highly appreciated!**

> [!NOTE]
>
> Please note that MonitorControl is no longer in active development. Translations submitted may make into a service/compatibility release eventually - but there is no guarantee. You might want to consider contributing to the localization of BetterDisplay instead. More info here: https://github.com/waydabber/BetterDisplay/discussions/2165

Here are some info regarding translation.

# Where are the files for the translation?

Translation files are located here: https://github.com/MonitorControl/MonitorControl/tree/master/MonitorControl/UI

If you see your language in the list (for example the folder `hu.lproj` contains the Hungarian translation), then you are in luck - you only need to edit the following files in the language folder via Github's own web interface to update an existing translation and then create a pull request:

- `Localizable.strings`
- `Main.strings`
- `InternetAccessPolicy.strings`

**IMPORTANT: All three files must be translated!**

If you do not see your language, then you have two options:

1. Either download Xcode and checkout the project, add a new language and then edit the source files (see the screenshot below) and then commit the changes and create a pull request (this method is more suitable for developers).
2. Simply make a copy of the English defaults [from here](https://github.com/MonitorControl/MonitorControl/tree/master/MonitorControl/UI/en.lproj) and complete the translation using your preferred text editor (this can be Xcode but you can use [BBEdit](https://apps.apple.com/us/app/bbedit/id404009241?mt=12) or any text editor you like) and attach the translated files to this discussion stating the language you want to add and the @MonitorControl/dev team will add it to the project.

<img width="1469" alt="Screen Shot 2021-10-01 at 11 03 21" src="https://user-images.githubusercontent.com/37590873/135595774-5192d590-9562-40cf-9933-daf07d688bd7.png">

# How to translate?

Making the translation is very simple. Just edit `Localizable.strings`, `Main.strings` and `InternetAccessPolicy.strings` and change the English text to the desired text in the target language.

- In the `Localizable.strings` and `InternetAccessPolicy.strings` files the text _after_ the equation sign should be translated, making sure that the quote signs remain intact and no quote signs are used inside the translated text:

<div align="center">
<img width="1168" alt="Screen Shot 2021-10-01 at 11 27 27" src="https://user-images.githubusercontent.com/37590873/135598009-3bc60845-65ac-425a-a939-e9248d692e68.png">
</div>

- In the `Main.strings` file you also need to change the text after the equation sign (the format is a little bit different).

<div align="center">
<img width="1168" alt="Screen Shot 2021-10-01 at 11 27 44" src="https://user-images.githubusercontent.com/37590873/135598027-c7600568-2a48-4fe8-8b05-945e1bb3704f.png">
</div>

The screenshots are from Xcode but you can use your own text editor as mentioned or Github's own editor. You can also review existing completed translation files to get the idea.

# How to translate the custom keyboard shortcut box?

The texts for the custom keyboard shortcuts are not available in MonitorControl's repository so these will be left out of the translation.

![135706661-2c8b6822-e56e-4b7e-9f36-49dff1d8d116](https://user-images.githubusercontent.com/37590873/135708910-e969b5bd-529b-40a7-bea7-a8cc6fca9cba.png)

The custom keyboard shortcuts are handled by a different repository, [KeyboardShortcuts](https://github.com/sindresorhus/KeyboardShortcuts), so localizations can be added there. **It is optional**, but since it is easy to do (only a few lines of text), we recommend contributing to that repo as well:

https://github.com/sindresorhus/KeyboardShortcuts/tree/main/Sources/KeyboardShortcuts/Localization

For more information on how to proceed with this, see this: https://github.com/sindresorhus/KeyboardShortcuts#localization

Once the localization is accepted by the owner of that repository, it will be automatically incorporated into MonitorControl as well upon the next version.

# How to try out my translation?

**This is optional!** You might want to build the app in Xcode (see [README.md](https://github.com/MonitorControl/MonitorControl#readme) for instructions) and build the app to see if your translation works well (text is not clipped, etc).

# Translation of version 4.0.0

- Translations from v3.0.0 was carried over but due to the number of interface changes over 80% of the interface must be retranslated.
- We will probably to remove incomplete translations from v4.0.0 final (but will add back them later when they are completed).
- Please be prepared that the translation work might take up about 1-2 hours due to the amount of text to be translated. **But keep in mind, you help a lot of users with this, your contribution and work makes a difference!**

## The current list of completed translations for v4.0.0 is the following:

- _English - base language_
- Chinese - thanks to @dev-coco, @anmoliyang
- Chinese (Traditional, Taiwan) - thanks to @stiiveo, @dororojames
- Czech - thanks to @isametry
- Dutch - thanks to @JoniVR 
- French - thanks to @the0neyouseek 
- German - thanks to @jajoho, @curana
- Hungarian - thanks to @waydabber
- Italian - thanks to @picov
- Korean - thanks to @zzulu
- Polish - thanks to @Zbynius
- Spanish - thanks to @OliverCG
- Turkish - thanks to @mennan, @ozkanozcan

**Comment by waydabber** [comment]
Thank you @jajoho for the German translation! https://github.com/MonitorControl/MonitorControl/pull/636

**Comment by JoniVR** [comment]
As a side note, it might be useful to move this awesome explanation into a separate markdown file and link it in the main README (that or move it to a Wiki page) as this seems valuable information outside of the v4.0.0 release too. Lots of people do translations without coding experience, so this is very helpful to them.

**Reply by waydabber** [reply]
Great idea! I liked this forum into the readme as the translation status for 4.0.0 is dynamically changing and the post is being updated from time-to-time with new info + translation status. Once the translations are mostly complete, we can separate the generic part of the explanation and add it to the repo as a separate markdown file (or we can use the wiki but that seems to be rather deserted :)).

**Comment by picov** [comment]
Italian translation from current master is here.
Tested in real app built with Xcode 13.
[it.lproj.zip](https://github.com/MonitorControl/MonitorControl/files/7271469/it.lproj.zip)

Thank you for this great app!

**Reply by waydabber** [reply]
Wonderful! Thank you! :)

**Reply by waydabber** [reply]
Here is the PR for the translation: https://github.com/MonitorControl/MonitorControl/pull/661

**Comment by anmoliyang** [comment]
Chinese translation is here.
[zh-Hans.lproj.zip](https://github.com/MonitorControl/MonitorControl/files/7314876/zh-Hans.lproj.zip)

**Reply by waydabber** [reply]
Thank you @anmoliyang!

I found a PR for this as well, I approved it: https://github.com/MonitorControl/MonitorControl/pull/690. If these are not the same, please feel free to improve the translation from the PR if needed. I noted both of you authors as translators in the opening post. If this is not right this way, please let me know!

**Comment by JoniVR** [comment]
@waydabber Another thing that I recently discovered (not sure if we knew this already), which may make translations easier:

In Xcode, click `Product > Export Localizations...`. This will download a folder with all localizable stuff. When you open the file for each language, a special editor will open. After editing and saving, you can then use the same menu to `Import Localizations...`

![image](https://user-images.githubusercontent.com/7591717/136658850-eff9b2c5-e1cc-4c28-9d01-5274c43013e9.png)

**Reply by waydabber** [reply]
Nice!

**Reply by kant** [reply]
@waydabber Minor proposal. In this current document there is a typo in line 03: `translataion` > `translation`.

**Reply by waydabber** [reply]
Fixed. :)

**Comment by OliverCG** [comment]
I've just made the commit and pull with the Spanish language, but just in case it was not made properly, here is. 😉 
[es.lproj.zip](https://github.com/MonitorControl/MonitorControl/files/7521931/es.lproj.zip)

**Reply by waydabber** [reply]
Hey @OliverCG - thanks for your contribution! I don't see the Pull Request somehow. Are you sure it went through? Thanks a lot! :)

**Reply by OliverCG** [reply]
Hi @waydabber, for some reason I can't Pull Request. Xcode is always asking for my credentials and I'm pretty sure I'm introducing the account details (user and token) properly. :-/ I have 2FA enabled, but I'm not sure if it's incompatible.

**Reply by waydabber** [reply]
Hey, no problem. I'll use the attached zip then and create the PR myself if you want - its just that this way you'll not get your contribution attributed to you by GitHub. Is that OK with you?

**Reply by OliverCG** [reply]
Let me try to fix it and I'll get back to you otherwise. 😉

**Reply by waydabber** [reply]
Ok! Let me know if you want me to do it! Thanks!

**Reply by OliverCG** [reply]
I could fix the problem with GitHub Desktop! 💪 Thanks again!

**Reply by waydabber** [reply]
Great, thank you, I approved it!

**Comment by gpnunes75** [comment]
Here is my 2 cents, PT-BR translation.

I don't have and don't know how to use Xcode so I can't test.

But I tried to keep all texts with same length. If you find any issue I will be glad to fix.

[pt-br.lproj.zip](https://github.com/MonitorControl/MonitorControl/files/7678205/pt-br.lproj.zip)
.

**Reply by waydabber** [reply]
Hey @gpnunes75 - amazing, thanks a lot! I'll double check it and if everything is fine, I'll create a PR for it! :)

**Reply by waydabber** [reply]
fyi: https://github.com/MonitorControl/MonitorControl/issues/838

**Reply by waydabber** [reply]
Here is the PR: https://github.com/MonitorControl/MonitorControl/pull/840

Thanks again @gpnunes75!

**Comment by jajoho** [comment]
@waydabber: it is not already easy to find all the strings that need to be translated, when text is added with a new version.
Is there a way to highlight them or put them in a list so that everyone who wants to translate the app knows what is new?

**Reply by waydabber** [reply]
Hi @jajoho, you can pull the project in XCode and compare the various commits to see what changed:

https://stackoverflow.com/questions/10666312/using-xcode-how-can-i-diff-between-a-working-copy-of-a-file-and-the-latest-comm/43076749

You can also use GitHub for the job:

https://docs.github.com/en/pull-requests/committing-changes-to-your-project/viewing-and-comparing-commits/comparing-commits

**Reply by jajoho** [reply]
@waydabber Thank you very much. Somehow I didn't think of it myself...

**Comment by isametry** [comment]
Hello, as per #958, here are my strings for the Czech translation (cs_CZ):

[cs.lproj.zip](https://github.com/MonitorControl/MonitorControl/files/8118403/cs.lproj.zip)

**Reply by waydabber** [reply]
Ok, thanks @isametry I'll add this one in a PR soon!

**Reply by waydabber** [reply]
PR added: https://github.com/MonitorControl/MonitorControl/pull/959

**Comment by Svec-Tomas** [comment]
Hey! Please find attached all 3 files translated to Slovak!
[Monitor Control Slovak.zip](https://github.com/MonitorControl/MonitorControl/files/13757456/Monitor.Control.Slovak.zip)

**Reply by Svec-Tomas** [reply]
Is all OK with my translation, please? Or is there anything else I need to do? Thanks.

**Reply by waydabber** [reply]
Hi there, thanks for the effort! Yes, a Pull Request should be added with the translation. I'll do it if you don't know how to, let me know!

**Reply by Svec-Tomas** [reply]
I am not really familiar with Github and I find it really confusing. But what I enjoy is making quality apps available in my native language. It would be great if you could do it, thanks in advance!

**Reply by waydabber** [reply]
All right, will do! I am not sure when or if there will be an updated release for MonitorControl though (development was ceased but there might be a service release maybe with updated translations - it's up to the team).

Note: if you are interested in translating [BetterDisplay](https://betterdisplay.pro) to Slovak, you can join the translation team - https://github.com/waydabber/BetterDisplay/discussions/2165 (I do provide a license for free for those who contribute meaningfully to the translation effort).

**Reply by waydabber** [reply]
I added a PR. An other team member will need to review and approve it. Thanks again!
